[CBRD-20340] Redesign the log page buffer.

funkydvd opened the pull request:
http://jira.cubrid.org/browse/CBRD-20340

Changes:
* Replaced hash with an array to store log pages in memory.
* Removed synchronization on reading log pages from buffer.
* Changed some return types.

ribram commented:
##### log_page_buffer.c

1 Remove PB_DATA_SIZE and use prm_get_integer_value (PRM_ID_LOG_NBUFFERS)

2 Remove log_pb_global_data.cursor and fst.

3 Remove SIZEOF_LOG_BUFFER and replace it with sizeof (LOG_BUFFER).

4 Remove LOG_GET_LOG_BUFFER_PTR (never used).

5 log_pages_area => we do not require an array of pointers to pages in log_pages_area.
  * Make log_Pb.pages_area LOG_PAGE *. On init, replace log_pages_area with log_Pb.pages_area. Do not allocate another array of LOG_PAGE **.
  * Same thing is valid for buffers. We can use LOG_BUFFER * directly, we don't need the array of pointers.

6 Rename log_get_log_buffer_ptr to logpb_get_log_buffer.
  * Add STATIC_INLINE prefix to logpb_get_log_buffer.
  * Declaration:
  ```c
  STATIC_INLINE LOG_BUFFER *logpb_get_log_buffer (LOG_PAGE * log_pg) __attribute__ ((ALWAYS_INLINE));
  ```
  * Definition:
  ```c
  STATIC_INLINE LOG_BUFFER *
  logpb_get_log_buffer (LOG_PAGE * log_pg)
  ```
  * Change function body to:
  ```c
  {
    int index;
  
    if (log_pg == log_Pb.header_page)
      {
        return log_Pb.header_buffer;
      }
    
    index = (log_pg - log_Pb.pages_area) / LOG_PAGESIZE;
    /* Safe guard: index is valid. */
    assert (index >= 0 && index < log_Pb.num_buffers);
    /* Safe guard: log_pg is correctly aligned. */
    assert (log_Pb.pages_area + LOG_PAGESIZE * index == log_pg);
    
    return log_Pb.buffers + index;
  }
  ```

7 Rename, change and make MOD function STATIC_INLINE (same as above)
  ```c
  int
  MOD (int x)
  {
    return x % PB_DATA_SIZE;
  }
  ```
  =>
  ```c
  STATIC_INLINE int
  log_get_log_buffer_index (LOG_PAGEID log_pageid)
  {
    return log_pageid % log_Pb.num_buffers;
  }
  ```

to be continued :).

ribram commented:
###### logpb_finalize_pool
log_Pb.pages_area must be freed.

ribram commented:
log_page_buffer.c

1 remove unused:
  LOGPB_FIND_NBUFFER_FROM_CONT_BUFFERS
  LOG_MAX_NUM_CONTIGUOUS_BUFFERS
  LOG_BKUP_HASH_NUM_PAGEIDS
  LOG_BUFAREA
  size var in logpb_initialize_pool
  log_Stat.log_buffer_expand_count
  
2 logpb_initialize_pool
  * Remove size
  -int i, size;
  +int i;

  * Add assert log_Pb.pages_area is NULL. Remove setting buffers to NULL.
  assert (log_Pb.buffers == NULL);
  +assert (log_Pb.pages_area == NULL);  
  -log_Pb.buffers = NULL;

  -size = ((log_Pb.num_buffers * sizeof (LOG_BUFFER)) + sizeof (LOG_BUFAREA));

  * Initialize pages_area once.
  -/* Initialize every new buffer */
  +/* Inititalize pages_area and buffers */
  +MEM_REGION_INIT (log_Pb.pages_area, (size_t) log_Pb.num_buffers * LOG_PAGESIZE);
  for (i = 0; i < log_Pb.num_buffers; i++)
    {
      -MEM_REGION_INIT (((char *) log_Pb.pages_area + i * (LOG_PAGESIZE)), (LOG_PAGESIZE));
      logpb_initialize_log_buffer (&log_Pb.buffers[i], (LOG_PAGE *) ((char *) log_Pb.pages_area + i * (LOG_PAGESIZE)));
    }

3 logpb_initialize_log_buffer
  * Remove comment "Scramble the content of buffers. ..."
   
4 logpb_invalidate_pool
  * Remove check for LOGPB_HEADER_PAGE_ID (cannot be in buffer).

  for (i = 0; i < log_Pb.num_buffers; i++)
    {
      log_bufptr = LOGPB_FIND_BUFPTR (i);
      -if ((log_bufptr->pageid == LOGPB_HEADER_PAGE_ID || log_bufptr->pageid > NULL_PAGEID) && log_bufptr->fcnt <= 0
	        -&& log_bufptr->dirty == false)
      +if (log_bufptr->pageid != NULL_PAGEID && log_bufptr->fcnt <= 0 && !log_bufptr->dirty == false)
	      {
	        logpb_initialize_log_buffer (log_bufptr, log_bufptr->logpage);
	      }
    }

5 error_code.h
  * Add comment to ER_LOG_NBUFFERS_TOO_SMALL /* obsolete */

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -412,58 +372,40 @@ static void logpb_append_crumbs (THREAD_ENTRY * thread_p, int num_crumbs, const
 static int logpb_copy_log_header (THREAD_ENTRY * thread_p, LOG_HEADER * to_hdr, const LOG_HEADER * from_hdr);
-
+STATIC_INLINE LOG_BUFFER *logpb_get_log_buffer (LOG_PAGE * log_pg) __attribute__ ((ALWAYS_INLINE));
+STATIC_INLINE int log_get_log_buffer_index (LOG_PAGEID log_pageid);
```
@funkydvd 

Please append `__attribute__ ((ALWAYS_INLINE))`.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -412,58 +372,40 @@ static void logpb_append_crumbs (THREAD_ENTRY * thread_p, int num_crumbs, const
- */
-static void
-logpb_move_next_clock_hand (void)
+LOG_BUFFER *
```
@funkydvd 

Please correct it as 
```C
STATIC_INLINE LOG_BUFFER *
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -412,58 +372,40 @@ static void logpb_append_crumbs (THREAD_ENTRY * thread_p, int num_crumbs, const
-      log_Pb.clock_hand = 0;
-    }
+
+  long long index = ((char *) log_pg - (char *) log_Pb.pages_area) / LOG_PAGESIZE;
```
@funkydvd 

It seems you suffered a warning. 
What about using `int` for `index` and casting the expression as `int`?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -412,58 +372,40 @@ static void logpb_append_crumbs (THREAD_ENTRY * thread_p, int num_crumbs, const
+  assert (index >= 0 && index < log_Pb.num_buffers);
+  /* Safe guard: log_pg is correctly aligned. */
+  assert ((char *) log_Pb.pages_area + LOG_PAGESIZE * index == (char *) log_pg);
+  return &log_Pb.buffers[index];
```
@funkydvd 

I'd like to propose 

```C
  if (log_pg == log_Pb.header_page)
    {
      return &log_Pb.header_buffer;
    }

  /* Safe guard: index is valid. */
  assert (index >= 0 && index < log_Pb.num_buffers);

  /* Safe guard: log_pg is correctly aligned. */
  assert ((char *) log_Pb.pages_area + (LOG_PAGESIZE * index) == (char *) log_pg);

  return &log_Pb.buffers[index];
```

Please let me remind you our coding convention. 
We always use explicit block even though it is a single statement. It prevents developers from committing unintended misses. 

Please also don't spare horizontal blank lines. It will greatly help developers to easily read codes.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -485,8 +427,6 @@ logpb_unfix_page (LOG_BUFFER * bufptr)
       er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_FREEING_TOO_MUCH, 0);
```
@funkydvd 

Please add an assertion as 
```C
  assert (bufptr->fcnt >= 0);
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -412,58 +372,40 @@ static void logpb_append_crumbs (THREAD_ENTRY * thread_p, int num_crumbs, const
-static void
-logpb_reset_clock_hand (int buffer_index)
+STATIC_INLINE int
+log_get_log_buffer_index (LOG_PAGEID log_pageid)
```
@funkydvd 

Would you rename it as `logpb_get_log_buffer_index`?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -726,35 +527,58 @@ logpb_initialize_pool (THREAD_ENTRY * thread_p)
+
+      free_and_init (log_Pb.buffers);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1,
+	      log_Pb.num_buffers * sizeof (*log_Pb.buffers));
```
@funkydvd 

Please correct the size parameter of the error message. It looks like copy-paste error. :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -726,35 +527,58 @@ logpb_initialize_pool (THREAD_ENTRY * thread_p)
+      logpb_initialize_log_buffer (&log_Pb.buffers[i], (LOG_PAGE *) ((char *) log_Pb.pages_area + i * (LOG_PAGESIZE)));
+    }
+
+  log_Pb.header_page = (LOG_PAGE *) malloc (LOG_PAGESIZE);
```
@funkydvd 

Please add error check for `log_Pb.header_page`.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,95 +807,64 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
 	{
-	  log_bufptr->logpage.hdr.logical_pageid = pageid;
-	  log_bufptr->logpage.hdr.offset = NULL_OFFSET;
+	  log_bufptr->fcnt++;
```
@funkydvd 

I'm not quite sure we can just replace the existing buffer. Please imagine a case that the existing page is dirty. 
You can refer to the existing code of `logpb_replace` to handle a dirty log buffer:
```C
                  if (log_bufptr->dirty == true)
                    {
                      END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);

                      assert (LOG_CS_OWN_WRITE_MODE (thread_p));
                      log_Stat.log_buffer_flush_count_by_replacement++;
                      logpb_flush_all_append_pages (thread_p);

                      START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
                      mnt_log_replacements (thread_p);
                      *retry = true;
                      return NULL;
                    }
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,95 +807,64 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
+	  log_bufptr->pageid = pageid;
+	  phy_pageid = logpb_to_physical_pageid (pageid);
+	  log_bufptr->phy_pageid = phy_pageid;
+	  return (log_bufptr->logpage);
```
@funkydvd 

Please also care monitoring stuffs:
```C
  if (is_perf_tracking)
    ...
```

ribram commented:
@funkydvd 

##### vacuum.c
Replace if (logpb_fetch_page () != NO_ERROR) with
```c
error = logpb_fetch_page ();
if (error != NO_ERROR)
  {
    ASSERT_ERROR ();
    ...
    [return error];  // if error is returned
  }
```

ASSERT_ERROR () macro makes sure that logpb_fetch_page has sets errors using er_set () function before returning error.

ribram commented:
@fanzaiqiang 

#### log_impl.h
Two empty lines at line 129. Please keep only one empty line.

ribram commented:
@funkydvd 

#### log_page_buffer.c

##### Descriptions
Add description comments to logpb_get_log_buffer_index, logpb_get_log_buffer.

Description comment format:
/*
 * func_name - describe functionality
 *
 * return : describe returned value
 * arg1 (in) : describe arg1
 * arg2 (in/out) : describe arg2
 * arg3 (out) : describe arg3
[*
 * Other notes or details. Optional.]
 */
 
##### logpb_get_log_buffer
Delay index index computation after early out:
```c
  int index;
  if (log_pg == log_Pb.header_page)
    {
      return &log_Pb.header_buffer;
    }
  index = (int) ((char *) log_pg - (char *) log_Pb.pages_area) / LOG_PAGESIZE;
```

##### logpb_initialize_log_buffer
Remove empty line
```c
  error_code = logpb_initialize_flush_info ();

  if (error_code != NO_ERROR)
    {
      goto error;
    }
```

##### logpb_Initialized
Add a global variable to know when page buffer is initialized. We cannot rely on log_Pb.buffers since it is not initialized from the start.
```c
static bool logpb_Initialized = false;
```
Set it to true in logpb_initialize_pool, to false in logpb_initialize_pool (or early out if it is already false), return it in logpb_is_pool_initialized.
Use assert (logpb_Initialized); when we expect it to be initialized (any function that accesses page buffer).

##### logpb_fix_page
NOTE in function description comment is no longer valid. Update it to the way buffer works now.

Always use brackets to make scopes clear in if/else, for, while, do/while clauses. Only acceptable exception is in MACRO definitions.
```c
  else
    {
      log_bufptr = &log_Pb.buffers[logpb_get_log_buffer_index ((int) pageid)];
    }
```

Replace code to avoid switching from buffer to logpage and again to buffer:
```c
      if (logpb_is_dirty (thread_p, (log_bufptr->logpage)))
	{
	  logpb_flush_page (thread_p, (log_bufptr->logpage), FREE);
	}
```
with
```c
      if (bufptr->dirty == true)
        {
          if (logpb_write_page_to_disk (thread_p, log_bufptr->logpage, log_bufptr->pageid) != NO_ERROR)
            {
              assert_release (false);
              return NULL;
            }
        }
```

Seems "stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;" should be used for reading from files or when looking for buffers to replace. Since we removed the replace mechanism, I think the code belongs to fetch_mode == OLD_PAGE branch. It is just a thought, I don't really know its meaning. Please investigate.

Move repeated code out of if/else:
```c
      if (fetch_mode == NEW_PAGE)
        {
          log_bufptr->logpage->hdr.logical_pageid = pageid;
          log_bufptr->logpage->hdr.offset = NULL_OFFSET;
        }
      else
        {
          /* stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT; */
          logpb_read_page_from_file (thread_p, pageid, LOG_CS_FORCE_USE, log_bufptr->logpage);
          mnt_log_fetch_ioreads (thread_p);
        }
    log_bufptr->fcnt++;
    phy_pageid = logpb_to_physical_pageid (pageid);
```

We might consider replacing fix count with fix_owner. Fix count does not really make sense, since we do not expect concurrent fixes. There can only be one fixer. We can keep fix_owner as a mean to make sure that there is only one owner. One issue to consider is if the same thread fixes the same buffer multiple times.

Remove block:
```c
assert (log_bufptr != NULL);
  /* for debugging in release mode */
  if (log_bufptr == NULL)
    {
      return NULL;
    }
```
log_bufptr was already used several times. We don't expect it to be NULL. The assert check should be added before using it for the first time.

Replace assert (((UINTPTR) (log_bufptr->logpage->area) % 8) == 0); with
```c
assert (ASSERT_ALIGN (log_bufptr->logpage->area, MAX_ALIGNMENT);
```
Basically it is the same code, but the purpose is more clear.

##### logpb_is_any_dirty
Remove duplicate empty lines. The code convention does not allow more than one successive empty line. There are many such cases. Please compare all your changes with develop and make sure unnecessary empty lines are removed.

##### logpb_flush_page
Remove commented END_EXCLUSIVE_ACCESS_LOG_PB.

##### logpb_free_without_mutex
Remove the function and move its content to logpb_free_page

##### logpb_get_page_id
Type in description comment (character 'b' was accidentally added)

##### logpb_dump
Remove "/* modified mutex with assertion of CS */" comment. Future readers will not understand what you mean.

##### logpb_fetch_page
```c
  rv = logpb_copy_page (thread_p, req_lsa->pageid, access_mode, log_pgptr);
  if (rv != NO_ERROR || log_pgptr == NULL)
    {
      return ER_FAILED;
    }
  return NO_ERROR;
```
Same code for logpb_copy_page_from_log_buffer, logpb_copy_page_from_file and logpb_copy_page.

##### logpb_read_page_from_file
Please use return ER_FAILED instead of ER_GENERIC_ERROR. ER_FAILED is the usually default error code when no specific code exists. ER_GENERIC_ERROR is usually used by er_set.
Same for logpb_write_page_to_disk.

##### revert unnecessary indentation changes:
```c
#if defined(CUBRID_DEBUG)
  er_log_debug (ARG_FILE_LINE,
		"log_next_append_page: new page id(%lld) delayed page id(%lld) "
		"total_append_page_count(%ld) total_delayed_page_count(%ld) "
		" num_toflush(%d) use_append_page_sec(%f) need_flush(%d) "
		"commit_count(%ld) total_commit_count(%ld)\n", (int) log_Stat.last_append_pageid,
		(int) log_Stat.last_delayed_pageid, log_Stat.total_append_page_count,
		log_Stat.total_delayed_page_count, flush_info->num_toflush, log_Stat.use_append_page_sec, need_flush,
		log_Stat.last_commit_count_while_using_a_page, log_Stat.total_commit_count_while_using_a_page);
#endif /* CUBRID_DEBUG */
```
```c
	  if (er_errid () == ER_IO_WRITE_OUT_OF_SPACE)
	    {
	      er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_WRITE_OUT_OF_SPACE, 4, bufptr->pageid,
		      phy_pageid, log_Name_active, log_Gl.hdr.db_logpagesize);
	    }
	  else
	    {
	      er_set_with_oserror (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_WRITE, 3, bufptr->pageid,
				   phy_pageid, log_Name_active);
	    }
```

```c
static int
prior_lsa_gen_2pc_prepare_record (THREAD_ENTRY * thread_p, LOG_PRIOR_NODE * node, int gtran_length,
				  char *gtran_data, int lock_length, char *lock_data)
```

```c
LOG_PRIOR_NODE *
prior_lsa_alloc_and_copy_crumbs (THREAD_ENTRY * thread_p, LOG_RECTYPE rec_type, LOG_RCVINDEX rcvindex,
				 LOG_DATA_ADDR * addr, const int num_ucrumbs,
				 const LOG_CRUMB * ucrumbs, const int num_rcrumbs, const LOG_CRUMB * rcrumbs)
```

```c
#if defined(CUBRID_DEBUG)
      if (logpb_get_page_id (log_Gl.append.delayed_free_log_pgptr) != log_Gl.append.prev_lsa.pageid)
	{
	  er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_WRONG_FORCE_DELAYED, 2,
		  log_Gl.append.prev_lsa.pageid, logpb_get_page_id (log_Gl.append.delayed_free_log_pgptr));
	  goto error;
	}
#endif /* CUBRID_DEBUG */
```

```c
LOG_PAGE *
logpb_fetch_from_archive (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_PAGE * log_pgptr,
			  int *ret_arv_num, LOG_ARV_HEADER * ret_arv_hdr, bool is_fatal)
```
There are several other cases, please compare with develop to find all of them.

##### vacuum_get_global_oldest_active_mvccid
This does not really belong to log page buffer, but here I noticed it. Change return code for vacuum_get_global_oldest_active_mvccid to MVCCID (this is actually the correct type). Remove casts to UINT64.

##### logpb_flush_all_append_pages
remove hold_lpb_cs

ribram commented:
@funkydvd 

### log_page_buffer.c

##### logpb_unfix_page
Let's change the debug code to:
```c
static void
logpb_unfix_page (LOG_BUFFER * bufptr)
{
  assert (bufptr->fown == thread_get_current_entry_index ());
  bufptr->fown = -1;
}
```

##### logpb_fix_page
Use all 120 columns to write comments - in this case the NOTE in description (use line breaks for different paragraphs only).

Change getting index to:
```c
  int thread_index = thread_get_current_entry_index ();
```

Check page is not fixed:
```c
assert (log_bufptr != NULL);
assert (log_bufptr->fown == -1);  /* Should not be fixed */
```

Move duplicated assignment log_bufptr->fown = idx; after if (log_bufptr->pageid == NULL_PAGEID)/else.

##### logpb_flush_page
Please remove "/*END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p); end without start- could be deleted */" comments.

##### logpb_get_page_id
```c
  assert (bufptr->fown == thread_get_current_entry_index ());
```

##### logpb_print_hash_entry
Remove unused function.

##### logpb_copy_page_from_file
```c
  if (log_pgptr != NULL)
    {
      rv = logpb_read_page_from_file (thread_p, pageid, LOG_CS_FORCE_USE, log_pgptr);
      if (rv != NO_ERROR || log_pgptr == NULL)
	{
          LOG_CS_EXIT (thread_p);
	  return ER_FAILED;
	}
    }
```

##### logpb_copy_page
Initialize rv = NO_ERROR; Do not return rv immediately when error occurs. Make sure rv is not NO_ERROR and call goto exit. This way critical section exit is also handled. At the end of exit label, return rv instead NO_ERROR.

##### logpb_write_page_to_disk
```c
  if (log_pgptr == NULL)
    {
      return ER_FAILED;
    }
```

##### logpb_flush_all_append_pages
```c
  assert (bufptr->fown == thread_get_current_entry_index ());
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +773,70 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
+    }
+  else
+    {
+      log_bufptr = &log_Pb.buffers[logpb_get_log_buffer_index ((int) pageid)];
```
@funkydvd 

I'd like to propose to change the line as safe as possible. We may have a bad log pageid for some cases, for instance, log page corruption, bad log record and so on. 

```C
  index = logpb_get_log_buffer_index ((int) pageid);
  if (index is in boundary)
    log_bufptr = ...;
  else
    er_set (...);
    return NULL;
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +773,70 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
 	{
-	  log_bufptr = logpb_replace (thread_p, &retry);
-	  if (log_bufptr == NULL)
+	  if (logpb_write_page_to_disk (thread_p, log_bufptr->logpage, log_bufptr->pageid) != NO_ERROR)
```
@funkydvd 

Please reset `dirty` flag when the page is written.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +773,70 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
-	      goto error;
-	    }
+	  stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+	  logpb_read_page_from_file (thread_p, pageid, LOG_CS_FORCE_USE, log_bufptr->logpage);
```
@funkydvd 

error handling is required.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1269,15 +863,7 @@ logpb_set_dirty (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, int free_page)
-  START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
-
   bufptr->dirty = true;
-  if (free_page == FREE)
```
@funkydvd 

Please notice that we have a case of `FREE` for `free_page`. 
Please find `logpb_end_append`.

funkydvd replied:
I have removed this parameter, because all it did was decreasing the fcnt, and we removed the fix count. I have updated the function, and now it doesn't have the free_page argument.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2050,41 +1538,48 @@ logpb_fetch_page (THREAD_ENTRY * thread_p, LOG_LSA * req_lsa, LOG_CS_ACCESS_MODE
-
-  return ret_pgptr;
+  rv = logpb_copy_page (thread_p, req_lsa->pageid, access_mode, log_pgptr);
+  if (rv != NO_ERROR || log_pgptr == NULL)
```
@funkydvd 

Is there a case of `log_pgtr == NULL`?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1088,31 +739,30 @@ logpb_create (THREAD_ENTRY * thread_p, LOG_PAGEID pageid)
   PERF_PAGE_MODE stat_page_found = PERF_PAGE_MODE_OLD_IN_BUFFER;
-
+  int idx = 1;
+#if defined (SERVER_MODE)
```
@funkydvd 

I hope to have a horizontal blank line between declarations and a statement. 
Another blank line at the end of `#endif` is also nice.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1256,11 +853,8 @@ void
   /* Get the address of the buffer from the page. */
-  bufptr = LOG_GET_LOG_BUFFER_PTR (log_pgptr);
-
+  bufptr = logpb_get_log_buffer (log_pgptr);
```
@funkydvd 

Please add a blank line to separate declaration and statement.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1293,18 +879,14 @@ static bool
-  START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
 
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+  bufptr = logpb_get_log_buffer (log_pgptr);
```
@funkydvd 

I also hope to have a blank line to separate assert for readability.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1355,22 +935,19 @@ logpb_is_any_fix (THREAD_ENTRY * thread_p)
   bool ret;
-
-  START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
```
@funkydvd 

Please also add a blank line. :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2050,41 +1538,48 @@ logpb_fetch_page (THREAD_ENTRY * thread_p, LOG_LSA * req_lsa, LOG_CS_ACCESS_MODE
 {
-  LOG_PAGE *ret_pgptr = NULL;
-
+  int rv;
```
@funkydvd 

Please also add a blank line at the end.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2050,41 +1538,48 @@ logpb_fetch_page (THREAD_ENTRY * thread_p, LOG_LSA * req_lsa, LOG_CS_ACCESS_MODE
 logpb_copy_page_from_file (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_PAGE * log_pgptr)
 {
-  LOG_PAGE *ret_pgptr = NULL;
+  int rv;
```
@funkydvd 

Please also add a blank line at the end.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2093,39 +1588,43 @@ logpb_copy_page_from_file (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_PAGE
   PERF_PAGE_MODE stat_page_found = PERF_PAGE_MODE_OLD_IN_BUFFER;
   bool log_csect_entered = false;
-
+  int rv = NO_ERROR;
```
@funkydvd 

Please also add a blank line at the end.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2141,30 +1640,63 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
     }
 
-  END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
+  log_bufptr = &log_Pb.buffers[logpb_get_log_buffer_index ((int) pageid)];
```
@funkydvd 

To check validity of `pageid` would be nice. I'd like to be conservative as possible.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2141,30 +1640,63 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
+    }
+  mnt_log_fetch_ioreads (thread_p);
+  stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+  if (log_pgptr == NULL)
```
@funkydvd 

Do you want to have `if` block?

funkydvd replied:
The race was resolved here. I want the "if" block, and this is exactly why, because I want to check if the buffer is not modified. I will put this if back, I forgot why was there. With this condition, there will be no race problems.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2380,10 +1912,13 @@ logpb_write_page_to_disk (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, LOG_PAG
     }
-
-  return log_pgptr;
+  if (log_pgptr == NULL)
```
@funkydvd 

I don't think this helps.

eseokoh commented:
@funkydvd @ribram 

It is better to carefully examine the effect of removing `logpb_Rmutex_log_pb` from `logpb_flush_all_append_pages`.
Please notice that it will DEMOTE and PROMOTE LOG_CS for HA case. 
I don't know the current code is safe or not. I wanna make it sure.

eseokoh commented on src/query/vacuum.c:
```diff
@@ -7428,10 +7435,10 @@ vacuum_notify_server_shutdown (void)
  * return : Global oldest active MVCCID.
  */
-VACUUM_LOG_BLOCKID
+MVCCID
```
@funkydvd 

Please also correct 
```C
    VACUUM_LOG_BLOCKID crt_oldest = vacuum_get_global_oldest_active_mvccid ();
```
of `logtb_get_oldest_active_mvccid`.

funkydvd replied:
We checked logpb_flush_all_append_pages, and we removed the logpb_Rmutex_log_pb because between demote and promote there are other active mutexes: writer_info->flush_start_mutex, flush_info->flush_mutex, writer_info->flush_wait_mutex, writer_info->flush_end_mutex, so there will be only one thread entering those critical functions (like logpb_set_dirty). If you want, I can do for debug reasons some checks to see that we have at least one mutex.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1663,33 +1682,33 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
+  else
+    {
+      er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, pageid);
+      return NULL;
```
@funkydvd 

`NULL` should be `ER_LOG_PAGE_CORRUPTED`.

funkydvd replied:
I have changed it. Do you have another observations or it is everything clear now?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +771,86 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
+	    }
+	  log_bufptr->dirty = false;
 	}
+      log_bufptr->pageid = NULL_PAGEID;	/* invalidate buffer */
```
@funkydvd 

Please add 
```C
  mnt_log_replacements (thread_p);
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1088,30 +739,27 @@ logpb_create (THREAD_ENTRY * thread_p, LOG_PAGEID pageid)
+ *       that page.
  */
 static LOG_PAGE *
 logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetch_mode)
```
@funkydvd @ribram 

I'd like to propose to rename the function, because we have removed `logpb_unfix_page`. Generally, fix-unfix, lock-unlock works together. 
I have thought about `logpb_locate_page` and `logpb_load_page`. I prefer to the first one but I don't really love it. Can you think of a better name? :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +771,86 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
 	{
-	  log_bufptr->logpage.hdr.logical_pageid = pageid;
-	  log_bufptr->logpage.hdr.offset = NULL_OFFSET;
+
```
@funkydvd 

Would you remove extra blank lines on line 813 and 808?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +771,86 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
+	    {
+	      assert_release (false);
+	      return NULL;
+	    }
```
@funkydvd 

Please add
```C
  mnt_log_iowrites (thread_p, 1);
```

eseokoh replied:
@funkydvd 

sorry I'd like to change my proposal. 
Please ignore the above comment and add it to `logpb_write_page_to_disk`. I'm going to also leave a comment of the function.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2380,10 +1931,10 @@ logpb_write_page_to_disk (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, LOG_PAG
       logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "logpb_write_page_to_disk");
-      return NULL;
+      return ER_FAILED;
     }
```
@funkydvd 

Please add 
```C
  mnt_log_iowrites (thread_p, 1);
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1248,18 +859,15 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
-logpb_set_dirty (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, int free_page)
+logpb_set_dirty (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr)
 {
   LOG_BUFFER *bufptr;		/* Log buffer associated with given page */
```
@funkydvd 

Would you add a blank line?

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1390,28 +984,21 @@ int
 logpb_flush_page (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, int free_page)
```
@funkydvd 

Please notice that `free_page` is now obsoleted.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1441,74 +1028,17 @@ logpb_flush_page (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, int free_page)
-    }
-
-  END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
+  /*END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p); end without start- could be deleted */
```
@funkydvd 

You can remove two commented lines.

eseokoh commented:
@funkydvd 

Please find `contrib/gdb_debugging_scripts/log_records.gdb` depends on the old implementation and also change the debugging functions. :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1992,20 +1497,20 @@ logpb_flush_header (THREAD_ENTRY * thread_p)
  * logpb_fetch_page - Fetch a exist_log page using local buffer
  *
- * return: Pointer to the page or NULL
+ * return: NO_ERROR if everything is ok, else ER_GENERIC_ERROR
```
@funkydvd 

Please change `ER_GENERIC_ERROR` with `ER_FAILED`.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2050,41 +1555,49 @@ logpb_fetch_page (THREAD_ENTRY * thread_p, LOG_LSA * req_lsa, LOG_CS_ACCESS_MODE
-  return ret_pgptr;
+  rv = logpb_copy_page (thread_p, req_lsa->pageid, access_mode, log_pgptr);
+  if (rv != NO_ERROR)
+    {
```
@funkydvd 

Please add 
```C
  ASSERT_ERROR ();
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2050,41 +1555,49 @@ logpb_fetch_page (THREAD_ENTRY * thread_p, LOG_LSA * req_lsa, LOG_CS_ACCESS_MODE
-  return ret_pgptr;
+  rv = logpb_copy_page (thread_p, pageid, LOG_CS_FORCE_USE, log_pgptr);
+  if (rv != NO_ERROR)
+    {
```
@funkydvd 

Please add
```C
  ASSERT_ERROR ();
```

eseokoh commented:
@funkydvd @ribram 

I'd like to ask you a couple of topics:

1. log fetch related monitoring stuffs: We have `log_num_fetch_ioreads`, `log_num_ioreads`, `log_num_fetches` and it may confuse us. I think the last two would be enough. What do you think of?
1. `log_Gl.append.delayed_free_log_pgptr` Please examine the related logics once more.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1525,9 +1055,8 @@ logpb_get_page_id (LOG_PAGE * log_pgptr)
 {
```
@funkydvd 

`logpb_get_page_id` could also be `STATIC_INLINE`.

eseokoh replied:
It's possible but impractical because usages are only debugging support functions. 
Please forget it.

eseokoh commented on contrib/gdb_debugging_scripts/log_records.gdb:
```diff
@@ -24,15 +24,8 @@ end
-    end
-  end
+  set $idx = $arg0 % log_Pb.num_buffers
+  set $arg1 = log_Pb.buffers[$idx]->logpage
```
@funkydvd 

I think to examine the pageid of the buffer matches with the given argument. 
Please look at the behavior of the old function.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2380,10 +1929,11 @@ logpb_write_page_to_disk (THREAD_ENTRY * thread_p, LOG_PAGE * log_pgptr, LOG_PAG
 
       logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "logpb_write_page_to_disk");
-      return NULL;
+      mnt_log_iowrites (thread_p, 1);
```
@funkydvd 

Please move the statement to outside of `if` block. This is an error handling block. :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -1123,123 +771,85 @@ logpb_fix_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, PAGE_FETCH_MODE fetc
-	{
-	  logpb_initialize_log_buffer (log_bufptr);
-	  log_bufptr = NULL;
+	  mnt_log_ioreads (thread_p);
```
@funkydvd 

Please notice that `logpb_read_page_from_file` counts the number of log reads and remove `mnt_log_ioreads` call from here. :)

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2141,30 +1657,63 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
+	      rv = ER_FAILED;
+	      goto exit;
+	    }
+	  mnt_log_ioreads (thread_p);
```
@funkydvd 

You can also remove `mnt_log_ioreads` call.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -2141,30 +1657,63 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
+      rv = ER_FAILED;
+      goto exit;
     }
+  mnt_log_ioreads (thread_p);
```
@funkydvd 

Please remove `mnt_log_ioreads` call.

eseokoh commented:
@funkydvd @ribram 

I think `logpb_copy_page` may not be safe due to potential race condition.
Please suppose that there are multiple threads to simultaneously execute the function with `access_mode == LOG_CS_SAFE_READER`. When they compete a buffer frame with different pages, to access the buffer may not be atomic. The existing code does not suffer from such race condition because of `logpb_Rmutex_log_pb`.

```C
  index = logpb_get_log_buffer_index ((int) pageid);
  if (index >= 0 && index < log_Pb.num_buffers)
    {
      log_bufptr = &log_Pb.buffers[index];
    }
  else
    {
      er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, pageid);
      return ER_LOG_PAGE_CORRUPTED;
    }
  if (log_bufptr->pageid == pageid)
    {
      /* Copy page from log_bufptr into log_pgptr */
      memcpy (log_pgptr, log_bufptr->logpage, LOG_PAGESIZE);
      goto exit;
    }
```

`if (log_bufptr->pageid == pageid)` and `memcpy` may not be atomic.

I don't know there are no other potential race conditions in your fix. We have to look into codes more.

eseokoh commented:
@funkydvd 

Please leave a comment when you reach your conclusion on the recent changes.

funkydvd commented:
We have found that delayed_free_log_pgptr was useful for flushing, when there was more than one page to flush, and we changed this to be consistent with the new implementation, using some validations of log_Gl.append.prev_lsa and  log_Gl.hdr.append_lsa. The variable was useful in the following situation: if we have more than one page to be flushed, it was added an end of log at the current delayed page, then all the pages are flushed, and then the end of record is removed. This was done for safety, if not all the pages are flushed and the operation is cancelled. 

We have kept this functionality, but because we don't guarantee that the pointer is kept in the log page buffer we don't use it anymore. Now we have to call logpb_locate_page to be sure that we refer to the page that we want to flush, and not to some old page that is in the buffer. Instead of keeping the delay pointer, we check if the pageid of  log_Gl.append.prev_lsa is the same with the pageid of log_Gl.hdr.append_lsa. If it is not, it means that there is more than one page to be flushed.

 I also deleted another two useless statistics (last_delayed_pageid and total_delayed_page_count), because there are no more needed after we removed the delayed_freee_log_pgptr.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -4479,13 +3962,14 @@ logpb_flush_all_append_pages (THREAD_ENTRY * thread_p)
 
-  if (log_Gl.append.delayed_free_log_pgptr != NULL)
+  if (log_Gl.append.prev_lsa.pageid != NULL_PAGEID && log_Gl.hdr.append_lsa.pageid != NULL_PAGEID &&
+      log_Gl.append.prev_lsa.pageid != log_Gl.hdr.append_lsa.pageid)
```
@funkydvd 

Please format as GNU C style:
```C
  if (log_Gl.append.prev_lsa.pageid != NULL_PAGEID && log_Gl.hdr.append_lsa.pageid != NULL_PAGEID 
      && log_Gl.append.prev_lsa.pageid != log_Gl.hdr.append_lsa.pageid)
```

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -4495,24 +3979,14 @@ logpb_flush_all_append_pages (THREAD_ENTRY * thread_p)
-#endif /* CUBRID_DEBUG */
-
-      tmp_eof = (LOG_RECORD_HEADER *) LOG_PREV_APPEND_PTR ();
+      first_append_log_page = logpb_locate_page (thread_p, log_Gl.append.prev_lsa.pageid, OLD_PAGE);
```
@funkydvd 

Please handle errors of `logpb_locate_page`.

eseokoh commented on src/transaction/log_page_buffer.c:
```diff
@@ -5625,17 +5079,9 @@ logpb_end_append (THREAD_ENTRY * thread_p, LOG_RECORD_HEADER * header)
-    }
+  if (!(log_Gl.append.prev_lsa.pageid != NULL_PAGEID && log_Gl.hdr.append_lsa.pageid != NULL_PAGEID &&
+	log_Gl.append.prev_lsa.pageid != log_Gl.hdr.append_lsa.pageid))
+    logpb_set_dirty (thread_p, log_Gl.append.log_pgptr);
```
@funkydvd 

Please rewrite it as follows:

```C
  if (log_Gl.append.prev_lsa.pageid == NULL_PAGEID || log_Gl.hdr.append_lsa.pageid == NULL_PAGEID 
      || log_Gl.append.prev_lsa.pageid == log_Gl.hdr.append_lsa.pageid)
    {
      logpb_set_dirty (thread_p, log_Gl.append.log_pgptr);
    }
```

eseokoh commented:
@funkydvd @ribram 

The patch looks mature to be merged. 
Would you merge it?
Thank you for your amazing work. :)
